modules/fedimint-ln-client/src/pay.rs

```diff
@@ -258,6 +287,40 @@ impl LightningPayFunded {
     async fn gateway_pay_invoice(
         gateway: LightningGateway,
         payload: PayInvoicePayload,
+    ) -> Result<String, GatewayPayError> {
+        // If we don't succeed to reach the gateway in 30 seconds we'd rather abort the
+        // payment so the user isn't surprised by a heavily delayed payment
+        let deadline = fedimint_core::time::now() + GATEWAY_API_TIMEOUT;
+
+        let mut last_error = None;
+        loop {
+            let duration_til_deadline = deadline
+                .duration_since(fedimint_core::time::now())
+                .unwrap_or_default();
+            select! {
+                response = Self::try_gateway_pay_invoice(gateway.clone(), payload.clone()) => {
+                    match response {
+                        Ok(preimage) => return Ok(preimage),
+                        Err(e) => {
+                            warn!("Error while trying to reach gateway: {}", e);
+                            last_error = Some(e);
+                            sleep(RETRY_DELAY).await;
+                        }
+                    }
+                },
+                () = sleep(duration_til_deadline) => {
```

Comment:
I'm not sure this abortion is correct, because we only abort on client side. It seems it's possible that the payment succeeds and we still returns a timeout

Reply:
Good catch!